### PR TITLE
Create Integration Tests for AddCommand #327

### PR DESCRIPTION
**Pull request by Zhiyuan-Amos**
Part of #327

### CODE REVIEW

**Comment by CanIHasReview** [comment]
# v1

@Zhiyuan-Amos submitted v1 for review.

* [v1 1/4] [Refactor LogicManagerTest](https://github.com/se-edu/addressbook-level4/pull/396/commits/23812fa3e531625a69cb81146b02e2a31742ccf2)
* [v1 2/4] [Extract helper methods for Command Tests](https://github.com/se-edu/addressbook-level4/pull/396/commits/2d283dc754c4a35f02fe3bf3fd7bd9c20c272de3)
* [v1 3/4] [Extract Integration Tests for AddCommand](https://github.com/se-edu/addressbook-level4/pull/396/commits/3f29894d746c4954d80eb645c887f5c3874bf6bf)
* [v1 4/4] [Add Integration Test for AddCommand](https://github.com/se-edu/addressbook-level4/pull/396/commits/cfeddb6f9c14eecaa7ed3236cc1605541062f22c)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/dc9fe96a38a2b5f9bb5bb4c05c3ff1db47b1effd...cfeddb6f9c14eecaa7ed3236cc1605541062f22c))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/396/1/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

File: src/test/java/seedu/address/logic/commands/CommandIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+import seedu.address.model.AddressBook;
+import seedu.address.model.Model;
+import seedu.address.model.ModelManager;
+import seedu.address.model.ReadOnlyAddressBook;
+import seedu.address.model.UserPrefs;
+import seedu.address.model.person.ReadOnlyPerson;
+import seedu.address.testutil.TypicalPersons;
+
+/**
+ * A Test class for command integration tests.
```
Comment (MightyCupcakes) [inline src/test/java/seedu/address/logic/commands/CommandIntegrationTest.java]: `A base test class for all command integration tests` ?

**Comment by MightyCupcakes** [comment]
[v1 1/4]

> Refactor LogicManagerTest

This commit title is too vague and misleading since the same commit also changes `EditCommandIntegrationTest`

[v1 2/4]

> Let’s extract some methods from EditCommandIntegrationTest into a base class CommandIntegrationTest.

Since the preceding paragraph talks about common methods, it will be good to reference it here to provide a better link:

> Let’s extract these common methods from EditCommandIntegrationTest into a base class CommandIntegrationTest.

**Comment by MightyCupcakes** [comment]
[v1 4/4]
The currently is not needed in the first paragraph as it should be implied.

File: src/test/java/seedu/address/logic/commands/CommandIntegrationTest.java (hunk 1 above)
Comment (Zhiyuan-Amos) [inline src/test/java/seedu/address/logic/commands/CommandIntegrationTest.java]: Your suggestion sounds right. I'm following the comment from `AddressBookGuiTest` tho :P But yup, I'll change it to your suggestion.

**Comment by Zhiyuan-Amos** [comment]
Thanks for your review @MightyCupcakes ! :)

**Comment by CanIHasReview** [comment]
# v2

@Zhiyuan-Amos submitted v2 for review.

* [v2 1/4] [Extract LogicManagerTest\.TestDataHelper\#adam to PersonUtil](https://github.com/se-edu/addressbook-level4/pull/396/commits/b930ccb0ed7a2213820e1f46bac18adffd21cd89)
* [v2 2/4] [Extract helper methods for Command Tests](https://github.com/se-edu/addressbook-level4/pull/396/commits/358e37843c0e47ab9e358524db543f38739ed64b)
* [v2 3/4] [Extract Integration Tests for AddCommand](https://github.com/se-edu/addressbook-level4/pull/396/commits/8cb91411321b6db90f45d0df75f8807de04a7340)
* [v2 4/4] [Add Integration Test for AddCommand](https://github.com/se-edu/addressbook-level4/pull/396/commits/f44b77dcb2044c576759685e3c3d9f9d6a1ef643)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/dc9fe96a38a2b5f9bb5bb4c05c3ff1db47b1effd...f44b77dcb2044c576759685e3c3d9f9d6a1ef643)) ([:chart_with_upwards_trend: Interdiff between v1 and v2](https://github.com/se-edu/addressbook-level4/compare/e121e4c9ad3a8e98156bba34874222c5c86c20ac~1...e121e4c9ad3a8e98156bba34874222c5c86c20ac))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/396/2/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

**Comment by damithc** [comment]
1. PR/issue titles need to be more specific. e.g. `Remove redundant add command tests from LogicManagerTest`
2. Similar to my comment in other similar PRs, lets `add missing tests` first and `remove redundant tests` later. While the two things look like they are linked, they are done at two different places and involve two different sets of concerns. I'd prefer if this PR is localized to one area of the code base. We should be able to add missing lower level tests without worrying about what is covered by higher level tests.

**Comment by Zhiyuan-Amos** [comment]
@damithc Is there a need to remove redundant tests later for this PR? Rationale being that LogicManagerTest is a massive class but LogicManager is a very small class; more or less taking any small subset of existing tests in LogicManagerTest will be able to successfully test all possible code paths. As such, removing AddCommand tests should be fine.

**Comment by Zhiyuan-Amos** [comment]
@yamgent 

> Not sure why it was mentioned as a new test, because we do have one testing for invalid values in `execute_add_invalidPersonData()`.

My intention is not exactly to transfer from LogicManagerTest (i think it isn't clear in my commit messages x.x). The `execute_add_invalidPersonData()` imo belongs to Parser tests whereby only parsing of the command is tested (even though currently it's doing integration test now). Whereas over here it's an integration test i.e it tests with the model. 

Which is why I didn't remove the corresponding test (cos even though it's the same test, it should be tested at the Parser level first), that's why it isn't exactly a 'transfer' since there's no remove :P

**Comment by yamgent** [comment]
@Zhiyuan-Amos Ok. Though as per earlier comment, I guess we won't be removing anything for the next iteration anyway.

**Comment by CanIHasReview** [comment]
# v3

@Zhiyuan-Amos submitted v3 for review.

* [v3 1/3] [Extract LogicManagerTest\.TestDataHelper\#adam to PersonUtil](https://github.com/se-edu/addressbook-level4/pull/396/commits/908a269556a65afbc5753e180eb98e490d385c7e)
* [v3 2/3] [Extract helper methods for Command Tests](https://github.com/se-edu/addressbook-level4/pull/396/commits/4812a29333376ea73c4a535972faf07f22e4307c)
* [v3 3/3] [Create integration tests for AddCommand](https://github.com/se-edu/addressbook-level4/pull/396/commits/e609c9ac12d50749efa1c643b9874b9383a6ee5a)

([:books: Archive](https://github.com/se-edu/addressbook-level4/compare/a5d9fcc4cce6b798663a565091e38782d140b7e7...e609c9ac12d50749efa1c643b9874b9383a6ee5a)) ([:chart_with_upwards_trend: Interdiff between v2 and v3](https://github.com/se-edu/addressbook-level4/compare/2407f5f35d95dadb59c5c74da20b61b2d07aa63f~1...2407f5f35d95dadb59c5c74da20b61b2d07aa63f))
<details>
<summary>Checkout this PR version locally</summary>

```shell
git fetch https://github.com/se-edu/addressbook-level4.git refs/pr/396/3/head:BRANCHNAME
```
where `BRANCHNAME` is the name of the local branch you want to fetch this PR to.

</details>

**Comment by damithc** [comment]
@chao1995 wrote the EditCommandIntegration test, so he may be able to review this PR.

File: src/test/java/seedu/address/logic/commands/AddCommandIntegrationTest.java [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+import seedu.address.testutil.PersonUtil;
+
+/**
+ * An integration test class that tests add command's interaction with the Model.
+ */
+public class AddCommandIntegrationTest extends CommandIntegrationTest {
+
+    @Test
+    public void execute_validCommand_succeeds() throws Exception {
+        Person toAdd = PersonUtil.adam();
```
Comment (chao1995) [inline src/test/java/seedu/address/logic/commands/AddCommandIntegrationTest.java]: change `toAdd` to `personNotInModel`?

File: src/test/java/seedu/address/logic/commands/AddCommandIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+
+        assertCommandSuccess(command, expectedMessage, expectedAddressBook,
+                expectedAddressBook.getPersonList());
+    }
+
+    @Test
+    public void execute_duplicatePerson_throwsCommandException() throws Exception {
+        Person toAdd = PersonUtil.adam();
+
+        model.addPerson(toAdd); // person already in internal address book
```
Comment (chao1995) [inline src/test/java/seedu/address/logic/commands/AddCommandIntegrationTest.java]: Alternatively, we can get a person that is already in the model (similar to what is done in the `EditCommandIntegrationTest`: `model.getFilteredPersonList().get(ZERO_BASED_INDEX_FIRST_PERSON)`).

`toAdd` can change to `personInModel`/`firstPersonInModel`

File: src/test/java/seedu/address/logic/commands/AddCommandIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+        String userInput = PersonUtil.getAddCommand(toAdd);
+        Command command = prepareCommand(userInput);
+
+        assertCommandFailure(command, AddCommand.MESSAGE_DUPLICATE_PERSON);
+    }
+
+    @Test
+    public void execute_invalidPersonData_throwsCommandException() throws Exception {
+        Command command = prepareCommand("add []\\[;] p/12345 e/valid@e.mail a/valid, address");
```
Comment (chao1995) [inline src/test/java/seedu/address/logic/commands/AddCommandIntegrationTest.java]: This test looks like a unit test for the `AddCommandParser`.

**Comment by pyokagan** [comment]
@Zhiyuan-Amos Why is this on hold?

**Comment by Zhiyuan-Amos** [comment]
@pyokagan Currently have a few other PRs on hand, also I plan to finish up all the tests for EditCommand first before beginning on this. :>

**Pull request closed without merging**